src/CustomMarshalers/src/System/Runtime/InteropServices/CustomMarshalers/EnumVariantViewOfEnumerator.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections;
+using System.Runtime.InteropServices.ComTypes;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    [Serializable]
+    internal class EnumVariantViewOfEnumerator : IEnumVARIANT, ICustomAdapter, ICustomQueryInterface
+    {
+        private static readonly Guid IID_IManagedObject = new Guid("C3FCC19E-A970-11d2-8B5A-00A0C9B7C9C4");
+
+        public EnumVariantViewOfEnumerator(IEnumerator enumerator)
+        {
+            if (enumerator is null)
+            {
+                throw new ArgumentNullException(nameof(enumerator));
+            }
+
+            Enumerator = enumerator;
+        }
+        internal IEnumerator Enumerator { get; }
+
+        public IEnumVARIANT Clone()
+        {
+            if (Enumerator is ICloneable clonable)
+            {
+                return new EnumVariantViewOfEnumerator(clonable.Clone() as IEnumerator);
+            }
+            else
+            {
+                throw new COMException(SR.Arg_EnumNotCloneable, HResults.E_FAIL);
+            }
+        }
+
+        public int Next(int celt, object[] rgVar, IntPtr pceltFetched)
+        {
+            int numElements = 0;
+
+            try
+            {
+                if (celt > 0 && rgVar == null)
+                {
+                    return Marshal.GetHRForException(new ArgumentException());
+                }
+
+                while ((numElements < celt) && Enumerator.MoveNext())
+                {
+                    rgVar[numElements++] = Enumerator.Current;
+                }
+
+                if (pceltFetched != IntPtr.Zero)
+                {
+                    Marshal.WriteInt32(pceltFetched, numElements);
+                }
+            }
+            catch (Exception e)
+            {
+                return Marshal.GetHRForException(e);
```

Comment:
If we decide to keep the return `int`, then return `e.HResult` instead.

Reply:
The `int` returns are from the definition of the `IEnumVARIANT` interface in the `System.Runtime.InteropServices.ComTypes` namespace. I could write my own ComImport for `IEnumVARIANT` but I figured it was probably a better idea to use the one we have to reduce duplication.

Reply:
Nope. If that is how the interface is defined, leave it. We should chat about `GetHRForException()` so you are aware of why - in most cases - it should be avoided.